p4-fusion/main.cc

```diff
@@ -218,14 +221,22 @@ int Main(int argc, char** argv)
 	// Commit procedure start
 	Timer commitTimer;
 
+	auto totalChanges = changes.size();
 	auto noMerge = arguments.GetNoMerge();
-	for (size_t i = 0; i < changes.size(); i++)
+	int i(0);
+	while (!changes.empty())
 	{
-		ChangeList& cl = changes.at(i);
-
 		// Ensure the files are downloaded before committing them to the repository
-		PRINT("Waiting for download of CL " << cl.number << " to complete")
-		cl.WaitForDownload();
+		PRINT("Waiting for download of CL " << changes.front().number << " to complete")
+		// First, wait until downloaded so the changelist is no longer referenced
+		// in worker threads.
+		changes.front().WaitForDownload();
+
+		// Now move the changelist and pop it off the queue.
+		// Once this iteration is over, it will be destructed
+		// and memory is freed.
+		ChangeList cl = std::move(changes.front());
```

Comment:
Not an actionable comment, but I found it to be an unfortunate subtlety that we need to wait explicitly for the workers to stop referencing the old memory address before we use std::move here.

Reply:
yes it's quite annoying - I think a rework using two queues could fix it where we only schedule changelists for processing once the download is done but I'll save that for a future one